File: clisoap2.pl
use Mojolicious::Lite;
use SOAP::Lite;
use Mojo::UserAgent;
use Mojo::Util qw(url_escape);

$ENV{PERL_LWP_SSL_VERIFY_HOSTNAME} = 0;

get '/clisoap2' => sub {
    my $c = shift;
    my $n = $c->param('n') // 0;

    # 1. Consumir el servicio SOAP
    my $soap = SOAP::Lite
        ->proxy('https://www.dataaccess.com/webservicesserver/NumberConversion.wso')
        ->ns('http://www.dataaccess.com/webservicesserver/', 'm');
    my $som = $soap->NumberToWords(SOAP::Data->name('ubiNum' => $n));
    my $texto_ingles = $som->result // '';

    # 2. Traducir el resultado de inglés a español usando la API de MyMemory
    my $ua = Mojo::UserAgent->new;
    my $url = "https://api.mymemory.translated.net/get?q=" . url_escape($texto_ingles) . "&langpair=en|es";
    my $tx = $ua->get($url);

    my $texto_espanol = $tx->result->json('/responseData/translatedText') // $texto_ingles;

    # 3. Retornar el resultado traducido
    return $c->render(text => lc($texto_espanol));
};

app->start;

File: clisoap2.js
const express = require('express');
const soap = require('soap');
const app = express();
const port = 3000;

const wsdlUrl = 'https://www.dataaccess.com/webservicesserver/NumberConversion.wso?WSDL';

app.get('/clisoap2', (req, res) => {
    const n = req.query.n || 0;

    soap.createClient(wsdlUrl, (err, client) => {
        if (err) return res.status(500).send(err.message);

        client.NumberToWords({ ubiNum: n }, async (err, result) => {
            if (err) return res.status(500).send(err.message);

            const textoIngles = result.NumberToWordsResult.trim();

            try {
                // Traducir el resultado de inglés a español usando la API pública de MyMemory
                const urlTraduccion = `https://api.mymemory.translated.net/get?q=${encodeURIComponent(textoIngles)}&langpair=en|es`;
                const respuestaApi = await fetch(urlTraduccion);
                const datosJson = await respuestaApi.json();

                const textoEspanol = datosJson.responseData.translatedText;

                // Retornar el resultado traducido en minúsculas
                res.send(textoEspanol.toLowerCase());
            } catch (error) {
                res.status(500).send("Error en el proceso de traducción: " + error.message);
            }
        });
    });
});

app.listen(port, () => {
    console.log(`Servidor de la Versión 2 corriendo en http://localhost:${port}`);
});
